core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1605,6 +1605,8 @@ private void ensureParanoidOff() {
   /** Simplifies a list of terms and combines them into an OR.
    * Modifies the list in place. */
   private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
+    // Auxiliary map to simplify cases like: X <> A OR X <> B => TRUE
+    final Map<RexNode, Comparison> notEqualsComparisonMap = new HashMap<>();
     for (int i = 0; i < terms.size(); i++) {
```

Comment:
Shouldn't we invoke the `#simplifyComparison` firstly for the `OR` operator operands ? And BTW, this hashMap cache would trigger the digest computation of `RexNode`, could we have another way to identify the equivalence for RexNode ?

Reply:
recursion is already done in the `simplifyOr` call; which have called `simplifyList`

Reply:
Yep, it did, thanks.

Reply:
Thanks for you comment @danny0405 
I don't quite understand the issue with RexNode `digest`, theoretically this field should be computed in the constructor ("Effectively final. Set in each sub-class constructor, and never re-set"). Besides, we must keep in mind that this map will effectively store just RexInputeRef and RexFieldAccess.